File: go/剑指offer/面试题35. 复杂链表的复制.go
package main

type Node struct {
	Val    int
	Next   *Node
	Random *Node
}

func copyRandomList(head *Node) *Node {
	if head == nil {
		return head
	}

	cur := head
	for cur != nil {
		cloned := &Node{cur.Val, nil, nil}
		cloned.Next = cur.Next
		cur.Next = cloned
		cur = cur.Next.Next
	}

	cur = head
	for cur != nil {
		cloned := cur.Next
		if cur.Random != nil {
			cloned.Random = cur.Random.Next
		}
		cur = cloned.Next
	}

	cur = head
	newHead, cloned := cur.Next, cur.Next
	for cur != nil {
		cur.Next = cloned.Next
		cur = cur.Next
		if cur != nil {
			cloned.Next = cur.Next
		}
		cloned = cloned.Next
	}
	return newHead
}

File: python/剑指offer/面试题35. 复杂链表的复制.py
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        if not head:
            return None
        cur = head
        while cur:
            node = Node(cur.val,None,None)
            node.next = cur.next
            cur.next = node
            cur = node.next
        cur = head
        while cur:
            node = cur.next
            node.random = cur.random.next if cur.random else None
            cur = node.next
        cur = head
        newhead = cur.next
        while cur:
            node = cur.next
            cur.next = node.next
            cur = cur.next
            if cur:
                node.next = cur.next
            else:
                node.next = None
        return newhead